### PR TITLE
Sydent typing, part 3

### PR DESCRIPTION
**Pull request by DMRobertson**
Last bit of #1110. Back references #1123 and #1124.

Tentatively scheduled for Dec 17th.

Should be previewable at https://pr1125--matrix-org-previews.netlify.app/blog/2021/12/17/type-coverage-for-sydent-evaluation

Preview: https://pr1125--matrix-org-previews.netlify.app

### CODE REVIEW

File: gatsby/content/blog/2021/12/2021-12-17-sydent-typing-3.mdx [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+
+```python
+>>> from typing import get_type_hints
+>>> def foo(x: int) -> str: pass
+...
+>>> get_type_hints(foo)
+{'x': <class 'int'>, 'return': <class 'str'>}
+```
+
+Programs and libraries are free to use these annotations at runtime as they see fit. The most well-known examples are probably [`dataclasses`](), [`attrs`](https://www.attrs.org/en/stable/) with [`auto_attribs=True`](https://www.attrs.org/en/stable/types.html) and [`Pydantic`](https://pydantic-docs.helpmanual.io/). I'd be interested to learn if anyone else is consuming annotations at runtime!
```
Comment (H-Shay) [inline gatsby/content/blog/2021/12/2021-12-17-sydent-typing-3.mdx:255]: Did you want to add a url to the `dataclasses` link?

File: gatsby/content/blog/2021/12/2021-12-17-sydent-typing-3.mdx [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+
+## Summary
+
+All in all, in a two-week sprint we were able to get Sydent's mypy coverage from a precision of 83% up to 94%. Our work would have spotted the [bytes-versus-strings bug](https://github.com/matrix-org/sydent/pull/415); we understand why [the missing await](https://github.com/matrix-org/sydent/pull/413) wasn't detected. We [fixed](https://github.com/matrix-org/sydent/issues/419) [other](https://github.com/matrix-org/sydent/issues/449) [small](https://github.com/matrix-org/sydent/issues/447) [bugs](https://github.com/matrix-org/sydent/issues/445) too as part of the process. As well as bugs, I've hopefully made the source code clearer for future readers (but that one is hard to quantify).
```
Comment (H-Shay) [inline gatsby/content/blog/2021/12/2021-12-17-sydent-typing-3.mdx:259]:
```suggestion
All in all, in a two-week sprint we were able to get Sydent's mypy coverage from a precision of 83% up to 94%. Our work would have spotted the [bytes-versus-strings bug](https://github.com/matrix-org/sydent/pull/415); we understand why [the missing await](https://github.com/matrix-org/sydent/pull/413) wasn't detected. We [fixed](https://github.com/matrix-org/sydent/issues/419) [other](https://github.com/matrix-org/sydent/issues/449) [small](https://github.com/matrix-org/sydent/issues/447) [bugs](https://github.com/matrix-org/sydent/issues/445) too as part of the process. As well as fix bugs, I've hopefully made the source code clearer for future readers (but that one is hard to quantify).
```

**Comment by H-Shay** [comment]
I said this on the other ones but I'll say it again: fantastic job! I wonder if you could get this published on mypy's blog, or the like-I am sure people will find it helpful so the more eyes on it the better. Thanks for taking the time to do such a thorough write-up, I definitely learned things.

File: gatsby/content/blog/2021/12/2021-12-17-sydent-typing-3.mdx (hunk 1 above)
Comment (DMRobertson) [reply gatsby/content/blog/2021/12/2021-12-17-sydent-typing-3.mdx:255]: I did, thank you!

File: gatsby/content/blog/2021/12/2021-12-17-sydent-typing-3.mdx [hunk 3]
```diff
@@ -0,0 +1,272 @@
...
+
+```
+
+### The meaning of precision
+
+There are two metrics I chose to focus on:
+- the proportion of "imprecise" lines across the project; I also used the complement, `precision = 100% - imprecision`, and
+- the proportion of expressions whose type is not `Any`.
+
+These are plotted in the graph at the top of this writeup. I could see that precision and the proportion of typed expressions were correlated, but I didn't understand how they differed. I couldn't see an explanation in the mypy docs, so I went [digging](https://github.com/python/mypy/blob/d807e097d142a88a48af93314d15ad87e41d2f19/mypy/stats.py) into the mypy source code. My understanding is as follows.
```
Comment (callahad) [inline gatsby/content/blog/2021/12/2021-12-17-sydent-typing-3.mdx:161]: The plot is in the first post, now. Link to it?

File: gatsby/content/blog/2021/12/2021-12-17-sydent-typing-3.mdx [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+
+It's worth highlighting the `typeshed` project, which maintains stubs for the standard library, plus popular third-party libraries. I [submitted a PR](https://github.com/python/typeshed/pull/6150) to add a single type hint—it was a very pleasant experience! Microsoft has an [incubator](https://github.com/microsoft/python-type-stubs) of sorts for stubs too.
+
+### Other typecheckers
+
+After the sprint to improve coverage, I spent a short amount of time trying the alternative type checkers out there. This is by no means an authoritative, exhaustive survey—just my quick notes.
```
Comment (callahad) [inline gatsby/content/blog/2021/12/2021-12-17-sydent-typing-3.mdx:185]: Might be worth explicitly noting that "Mypy isn't the *only* typechecker in the Python ecosystem. Other companies have built their own tools to reason about type hints in Python, each with their own strengths and weaknesses."

File: gatsby/content/blog/2021/12/2021-12-17-sydent-typing-3.mdx [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+
+There's room to spin out contributions upstream too. I submitted [two](https://github.com/twisted/twisted/pull/1669) [PRs](https://github.com/twisted/twisted/pull/1668) to twisted upstream; have started to work on annotations for [pynacl](https://github.com/pyca/pynacl/pull/693) [in my](https://github.com/pyca/pynacl/pull/694) [spare time](https://github.com/pyca/pynacl/pull/695); and submitted a quick fix to [typeshed](https://github.com/python/typeshed/pull/6150).
+
+Looking forward, I think we'd get a quick gain from ensuring that our smaller libraries ([signedjson](https://github.com/matrix-org/python-signedjson), [canonicaljson](https://github.com/matrix-org/python-canonicaljson)) are annotated. We'll be sticking with mypy for now—`mypy-zope` is crucial given how much we make use of twisted. We've started to spent time on similar efforts to improve Sygnal and Synapse in the coming weeks—though not to the extreme standard of `--strict` across everything.
```
Comment (callahad) [inline gatsby/content/blog/2021/12/2021-12-17-sydent-typing-3.mdx:262]:
```suggestion
Looking forward, I think we'd get a quick gain from ensuring that our smaller libraries ([signedjson](https://github.com/matrix-org/python-signedjson), [canonicaljson](https://github.com/matrix-org/python-canonicaljson)) are annotated. We'll be sticking with mypy for now—the `mypy-zope` plugin is crucial given our reliance on twisted. We're also working to improve Sygnal and Synapse—though not to the extreme standard of `--strict` across everything.
```

File: gatsby/content/blog/2021/12/2021-12-17-sydent-typing-3.mdx (hunk 4 above)
Comment (DMRobertson) [reply gatsby/content/blog/2021/12/2021-12-17-sydent-typing-3.mdx]: Thanks, d9f8bcc

File: gatsby/content/blog/2021/12/2021-12-17-sydent-typing-3.mdx [hunk 6]
```diff
@@ -0,0 +1,272 @@
...
+
+```python
+>>> from typing import get_type_hints
+>>> def foo(x: int) -> str: pass
+...
+>>> get_type_hints(foo)
+{'x': <class 'int'>, 'return': <class 'str'>}
+```
+
+Programs and libraries are free to use these annotations at runtime as they see fit. The most well-known examples are probably ]dataclasses](https://docs.python.org/3/library/dataclasses.html), [attrs](https://www.attrs.org/en/stable/) with [auto_attribs=True](https://www.attrs.org/en/stable/types.html) and [Pydantic](https://pydantic-docs.helpmanual.io/). I'd be interested to learn if anyone else is consuming annotations at runtime!
```
[the remark this answers is not in the document]
Comment (DMRobertson) [reply gatsby/content/blog/2021/12/2021-12-17-sydent-typing-3.mdx]: Argh. Thanks. 7779f02

**Pull request merged**